File: crates/amux-tui/src/app/workspace_actor_picker.rs
use amux_protocol::{WorkspaceActor, AGENT_ID_SWAROG};

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub(super) enum WorkspaceActorPickerMode {
    Assignee,
    Reviewer,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub(super) struct WorkspaceActorPickerOption {
    pub(super) label: String,
    pub(super) actor: Option<WorkspaceActor>,
}

impl WorkspaceActorPickerMode {
    pub(super) fn title(self) -> &'static str {
        match self {
            Self::Assignee => "WORKSPACE ASSIGNEE",
            Self::Reviewer => "WORKSPACE REVIEWER",
        }
    }

    fn noun(self) -> &'static str {
        match self {
            Self::Assignee => "assignee",
            Self::Reviewer => "reviewer",
        }
    }
}

pub(super) fn workspace_actor_picker_options(
    mode: WorkspaceActorPickerMode,
    subagents: &crate::state::SubAgentsState,
) -> Vec<WorkspaceActorPickerOption> {
    let mut options = vec![WorkspaceActorPickerOption {
        label: "none".to_string(),
        actor: None,
    }];
    if mode == WorkspaceActorPickerMode::Reviewer {
        options.push(WorkspaceActorPickerOption {
            label: "user".to_string(),
            actor: Some(WorkspaceActor::User),
        });
    }
    options.push(WorkspaceActorPickerOption {
        label: "svarog".to_string(),
        actor: Some(WorkspaceActor::Agent(AGENT_ID_SWAROG.to_string())),
    });
    options.extend(builtin_workspace_actor_options());
    options.extend(
        subagents
            .entries
            .iter()
            .filter(|entry| entry.enabled)
            .map(|entry| WorkspaceActorPickerOption {
                label: if entry.name.trim().is_empty() {
                    entry.id.clone()
                } else {
                    entry.name.clone()
                },
                actor: Some(WorkspaceActor::Subagent(entry.id.clone())),
            }),
    );
    options
}

fn builtin_workspace_actor_options() -> impl Iterator<Item = WorkspaceActorPickerOption> {
    crate::state::subagents::BUILTIN_PERSONA_ROLE_CHOICES
        .iter()
        .filter(|choice| {
            !matches!(
                choice.id,
                amux_protocol::AGENT_ID_SWAROG | amux_protocol::AGENT_ID_RAROG | "rod"
            )
        })
        .map(|choice| WorkspaceActorPickerOption {
            label: choice.label.to_string(),
            actor: Some(WorkspaceActor::Subagent(choice.id.to_string())),
        })
}

pub(super) fn workspace_actor_picker_body(
    task_id: &str,
    mode: WorkspaceActorPickerMode,
    options: &[WorkspaceActorPickerOption],
    cursor: usize,
) -> String {
    let short_id = task_id.chars().take(12).collect::<String>();
    let mut body = format!("Task: {short_id}\nSelect {}:\n\n", mode.noun());
    for (index, option) in options.iter().enumerate() {
        let marker = if index == cursor { ">" } else { " " };
        body.push_str(&format!("{marker} {}\n", option.label));
    }
    body.push_str("\nEnter select - Esc cancel");
    body
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::state::{SubAgentEntry, SubAgentsState};

    fn subagent(id: &str, name: &str, enabled: bool) -> SubAgentEntry {
        SubAgentEntry {
            id: id.to_string(),
            name: name.to_string(),
            provider: "openai".to_string(),
            model: "gpt-5.4".to_string(),
            role: None,
            enabled,
            builtin: false,
            immutable_identity: false,
            disable_allowed: true,
            delete_allowed: true,
            protected_reason: None,
            reasoning_effort: None,
            raw_json: None,
        }
    }

    #[test]
    fn assignee_picker_excludes_user_and_disabled_subagents() {
        let mut subagents = SubAgentsState::new();
        subagents.entries = vec![subagent("qa", "QA", true), subagent("off", "Off", false)];

        let options =
            workspace_actor_picker_options(WorkspaceActorPickerMode::Assignee, &subagents);
        let labels = options
            .iter()
            .map(|option| option.label.as_str())
            .collect::<Vec<_>>();

        assert!(labels.starts_with(&["none", "svarog", "Weles"]));
        assert!(labels.contains(&"Mokosh"));
        assert!(labels.contains(&"QA"));
        assert!(!labels.contains(&"Off"));
    }

    #[test]
    fn reviewer_picker_includes_user() {
        let subagents = SubAgentsState::new();
        let options =
            workspace_actor_picker_options(WorkspaceActorPickerMode::Reviewer, &subagents);

        assert_eq!(options[0].label, "none");
        assert_eq!(options[1].label, "user");
        assert_eq!(options[2].label, "svarog");
        assert!(options.iter().any(|option| option.label == "Mokosh"
            && option.actor == Some(WorkspaceActor::Subagent("mokosh".to_string()))));
    }

    #[test]
    fn picker_body_marks_selected_option() {
        let subagents = SubAgentsState::new();
        let options =
            workspace_actor_picker_options(WorkspaceActorPickerMode::Reviewer, &subagents);

        let body = workspace_actor_picker_body(
            "workspace-task-123",
            WorkspaceActorPickerMode::Reviewer,
            &options,
            1,
        );

        assert!(body.contains("Task: workspace-ta"));
        assert!(body.contains("> user"));
    }
}

File: frontend/src/zorai/features/workspaces/workspaceActorPicker.ts
import type { SubAgentDefinition } from "@/lib/agentStore/types";

export type WorkspaceActorPickerMode = "assignee" | "reviewer";

export type WorkspaceActorPickerOption = {
  label: string;
  value: string;
  targetAgentId: string | null;
  requiresRuntimeSetup: boolean;
};

export const BUILTIN_WORKSPACE_PERSONAS = [
  { id: "weles", label: "Weles" },
  { id: "swarozyc", label: "Swarozyc" },
  { id: "radogost", label: "Radogost" },
  { id: "domowoj", label: "Domowoj" },
  { id: "swietowit", label: "Swietowit" },
  { id: "perun", label: "Perun" },
  { id: "mokosh", label: "Mokosh" },
  { id: "dazhbog", label: "Dazhbog" },
] as const;

export function workspaceActorPickerOptions(
  mode: WorkspaceActorPickerMode,
  subAgents: SubAgentDefinition[],
): WorkspaceActorPickerOption[] {
  const options: WorkspaceActorPickerOption[] = [
    { label: "none", value: "", targetAgentId: null, requiresRuntimeSetup: false },
  ];
  if (mode === "reviewer") {
    options.push({ label: "user", value: "user", targetAgentId: null, requiresRuntimeSetup: false });
  }
  options.push({ label: "svarog", value: "agent:svarog", targetAgentId: "svarog", requiresRuntimeSetup: false });
  options.push(
    ...BUILTIN_WORKSPACE_PERSONAS.map((persona) => {
      const existing = findSubAgent(subAgents, persona.id);
      return {
        label: persona.label,
        value: `subagent:${persona.id}`,
        targetAgentId: persona.id,
        requiresRuntimeSetup: needsBuiltinRuntimeSetup(existing),
      };
    }),
  );
  options.push(
    ...subAgents
      .filter((entry) => entry.enabled)
      .map((entry) => ({
        label: entry.name.trim() || entry.id,
        value: `subagent:${entry.id}`,
        targetAgentId: entry.id,
        requiresRuntimeSetup: needsBuiltinRuntimeSetup(entry),
      })),
  );
  return dedupeOptions(options);
}

export function workspaceActorOptionByValue(
  mode: WorkspaceActorPickerMode,
  subAgents: SubAgentDefinition[],
  value: string,
): WorkspaceActorPickerOption | null {
  const normalized = normalizeActorValue(value);
  return workspaceActorPickerOptions(mode, subAgents).find((option) => option.value === normalized) ?? null;
}

export function normalizeActorValue(value: string): string {
  const trimmed = value.trim();
  if (!trimmed || trimmed === "none") return "";
  if (trimmed === "user") return "user";
  if (trimmed === "svarog") return "agent:svarog";
  return trimmed;
}

function findSubAgent(subAgents: SubAgentDefinition[], id: string): SubAgentDefinition | null {
  const normalized = id.trim().toLowerCase();
  return subAgents.find((entry) => entry.id.trim().toLowerCase() === normalized) ?? null;
}

function needsBuiltinRuntimeSetup(entry: SubAgentDefinition | null | undefined): boolean {
  if (!entry) return true;
  return Boolean(entry.builtin && (!entry.provider.trim() || !entry.model.trim()));
}

function dedupeOptions(options: WorkspaceActorPickerOption[]): WorkspaceActorPickerOption[] {
  const seen = new Set<string>();
  return options.filter((option) => {
    if (seen.has(option.value)) return false;
    seen.add(option.value);
    return true;
  });
}
